[WIP] htlcswitch: use fn.GoroutineManager

starius opened the pull request:
## Change Description

Replaced the use of `s.quit` and `s.wg` with `s.gm` (`GoroutineManager`). WaitGroup is still needed to wait for `handleLocalResponse`: if it was switched to `s.gm`, then it may skip running, which has unclear consequences. After `handleLocalResponse` is changed to run without a goroutine, we can remove WaitGroup completely.

This fixes a race condition between `s.wg.Add(1)` and `s.wg.Wait()`. 

## Steps to Test

I added a test which used to fail under `-race` before this commit.

```
$ cd htlcswitch

$ go test -race -run TestSwitchGetAttemptResultStress
```

This test crashes with a data race if I undo the changes of implementation of switch.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

ellemouton commented:
@starius  - I think these unit test failures are related to this PR - maybe take a look at fixing those up first & then re-ping reviewers when ready?

lightninglabs-deploy commented:
@crypt-iq: review reminder
@starius, remember to re-request review from reviewers when ready

Crypt-iQ commented on htlcswitch/switch.go:
```diff
@@ -524,7 +528,12 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
+	})
+
+	// The switch shutting down is signaled by closing the channel.
+	if err != nil {
```
I think the switch shutting down and an error from the goroutine manager are different?

starius replied:
GoroutineManager can only return an error, if it is stopping. I added a check just in case:

```go
        // The switch shutting down is signaled by closing the channel.
        if errors.Is(err, fn.ErrStopping) {
                close(resultChan)
        } else if err != nil {
                return nil, fmt.Errorf("got an unexpected error from "+
                        "GoroutineManager: %w", err)
        }
```

starius replied:
This is now fixed in new GoroutineManager version. Method `Go` now returns `bool` (success). Resolving the thread.

Crypt-iQ commented on htlcswitch/switch.go:
```diff
@@ -1627,15 +1643,16 @@ out:
-			}()
+			})
+			if err != nil {
+				return
```
don't think this should return?

starius replied:
Fixed, added a comment. Now this section looks like this:
```go
                // When this time ticks, then it indicates that we should
                // collect all the forwarding events since the last internal,
                // and write them out to our log.
                case <-s.cfg.FwdEventTicker.Ticks():
                        // The error of Go is ignored: if it is shutting down,
                        // the loop will terminate on the next iteration, in
                        // s.gm.Done case.
                        _ = s.gm.Go(func(ctx context.Context) {
                                err := s.FlushForwardingEvents()
                                if err != nil {
                                        log.Errorf("unable to flush "+
                                                "forwarding events: %v", err)
                                }
                        })
```

Crypt-iQ reviewed (commented):
I couldn't reproduce the race condition with the attached test, do you have an error trace of it?

starius commented:
@Crypt-iQ 

> I couldn't reproduce the race condition with the attached test, do you have an error trace of it?

I pushed branch [reproduce-race](https://github.com/starius/lnd/tree/reproduce-race) to my fork.

In that branch:

```
htlcswitch$ go test -race -run TestSwitchGetAttemptResultStress
==================
WARNING: DATA RACE
Read at 0x00c000020118 by goroutine 21:
  runtime.raceread()
      <autogenerated>:1 +0x1e
  github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetAttemptResult()
      /home/user/lnd/htlcswitch/switch.go:496 +0x1c4
  github.com/lightningnetwork/lnd/htlcswitch.TestSwitchGetAttemptResultStress.func2()
      /home/user/lnd/htlcswitch/switch_test.go:3216 +0x168

Previous write at 0x00c000020118 by goroutine 22:
  runtime.racewrite()
      <autogenerated>:1 +0x1e
  github.com/lightningnetwork/lnd/htlcswitch.(*Switch).Stop()
      /home/user/lnd/htlcswitch/switch.go:1995 +0x1e9
  github.com/lightningnetwork/lnd/htlcswitch.TestSwitchGetAttemptResultStress.func3()
      /home/user/lnd/htlcswitch/switch_test.go:3230 +0xae

Goroutine 21 (running) created at:
  github.com/lightningnetwork/lnd/htlcswitch.TestSwitchGetAttemptResultStress()
      /home/user/lnd/htlcswitch/switch_test.go:3208 +0x45c
  testing.tRunner()
      /home/user/.goroot/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /home/user/.goroot/src/testing/testing.go:1743 +0x44

Goroutine 22 (finished) created at:
  github.com/lightningnetwork/lnd/htlcswitch.TestSwitchGetAttemptResultStress()
      /home/user/lnd/htlcswitch/switch_test.go:3225 +0x57a
  testing.tRunner()
      /home/user/.goroot/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /home/user/.goroot/src/testing/testing.go:1743 +0x44
==================
--- FAIL: TestSwitchGetAttemptResultStress (0.07s)
    switch_test.go:3197: 
                Error Trace:    /home/user/lnd/htlcswitch/switch_test.go:3197
                                                        /home/user/lnd/htlcswitch/switch_test.go:3234
                Error:          Received unexpected error:
                                htlc switch already shutdown
                Test:           TestSwitchGetAttemptResultStress
    testing.go:1399: race detected during execution of test
FAIL
exit status 1
FAIL    github.com/lightningnetwork/lnd/htlcswitch      0.395s
```

starius commented on htlcswitch/switch.go:
```diff
@@ -524,7 +528,12 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 		resultChan <- result
-	}()
+	})
+
```
Fixed. Thanks!

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3181,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
 	}
 }
 
+func TestSwitchGetAttemptResultStress(t *testing.T) {
```
Fixed:

```go
// TestSwitchGetAttemptResultStress runs series of GetAttemptResult and Stop in
// parallel to make sure there is no race condition between these actions.
func TestSwitchGetAttemptResultStress(t *testing.T) {
```

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3181,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
+	s, err := initSwitchWithTempDB(t, testStartingHeight)
+	require.NoError(t, err, "unable to init switch")
+	if err := s.Start(); err != nil {
+		t.Fatalf("unable to start switch: %v", err)
```
Fixed:

```go
require.NoError(t, s.Start(), "unable to start switch")
```

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3181,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
+	}
+	defer func() {
+		require.NoError(t, s.Stop())
+	}()
```
I removed this `defer`, since it is stopped during the test already, so another .Stop() call failed with "htlc switch already shutdown".

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3181,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
+	go func() {
+		defer wg.Done()
+
+		time.Sleep(10 * time.Millisecond)
```
This is by design of the test, to make s.Stop() call in parallel with GetAttemptResult calls. I added comments to the code of the test to clarify this:

```go
        // Run s.Stop() in parallel with consecutive GetAttemptResult calls to
        // make sure this doesn't result in a race condition.
        wg.Add(1)
        go func() {
                defer wg.Done()

                // Sleep 10ms to let several GetAttemptResult calls happen, so
                // s.Stop() happens in the middle of GetAttemptResult series.
                // The value 10ms was found empirically - this time is needed
                // to expose the race condition (as a crash under -race) in the
                // version of Switch before GoroutineManager was added.
                time.Sleep(10 * time.Millisecond)

                require.NoError(t, s.Stop())
        }()
```

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3183,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
+	const paymentID = 123
+	var preimg lntypes.Preimage
+	preimg[0] = 3
+
```
Great catch! Fixed!

starius commented on htlcswitch/switch_test.go:
```diff
@@ -3180,6 +3183,58 @@ func TestSwitchGetAttemptResult(t *testing.T) {
+
+		time.Sleep(10 * time.Millisecond)
+
+		require.NoError(t, s.Stop())
```
I removed the cleanup. The idea of this test is to call s.Stop() from a goroutine, in parallel to GetAttemptResult, to catch the race condition (if it is present).

starius commented:
> @starius - I think these unit test failures are related to this PR - maybe take a look at fixing those up first & then re-ping reviewers when ready?

Test failure was caused by extra call to s.Stop in `defer`. I removed it.

ellemouton commented on htlcswitch/switch_test.go:
```diff
@@ -3226,7 +3228,7 @@ func TestSwitchGetAttemptResultStress(t *testing.T) {
 		// The value 10ms was found empirically - this time is needed
 		// to expose the race condition (as a crash under -race) in the
-		// unfixed version of Switch, before GoroutineManager was added.
+		// version of Switch before GoroutineManager was added.
```
should be fixed in the previous commit that added it

starius replied:
Fixed. Thanks!

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -493,14 +500,11 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 		select {
 		case n = <-nChan:
-		case <-s.quit:
+		case <-ctx.Done():
```
this will never happen cause we pass it a fresh `context.Background()` on instantiation. 

This makes me think that the `Go(func(ctx context.Context))` signature should actually be: `Go(func(ctx context.Context, quit chan struct{}))` where `quit` is a channel that is closed when the goroutine manager is exiting. That way we can handle that case inside here. How we handle the exiting of the goroutine manager is thus more explicit.

ellemouton replied:
the reason I'm pointing this out is cause we now have 2 places calling `close` on the `resultChan` which is a red flag to me (due to the potential of a panic). It takes introspection of the `Go` call to see that these two can never be called simultaneously, but I dont think a caller should need to depend on that the internals of the helper

ellemouton replied:
> this will never happen cause we pass it a fresh context.Background() on instantiation.

ah ok I see we do create a child context off of it and call cancel on that if GoroutingManager is stopped. 

I think we should still unify this though: explicit context passed in and an explicit quit channel called when GotroutineManager is Stopped. Ie, dont use the same derived context for 2 things. That way we dont have to handle the ErrStopping in 2 places as is done here.

starius replied:
@ellemouton and me discussed the interface of `GoroutineManager.Go` method.
We are leaning towards this:
```go
Go(func(ctx context.Context)) bool

ok := Go(...)
if !ok {
  // Failed to launch a goroutine.
}
```

The quit channel is not needed, since the ctx is a child context which is cancelled upon goroutine manager stop.

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -524,7 +528,14 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 		}
 		resultChan <- result
-	}()
+	})
+	// The switch shutting down is signaled by closing the channel.
+	if errors.Is(err, fn.ErrStopping) {
+		close(resultChan)
+	} else if err != nil {
+		return nil, fmt.Errorf("got an unexpected error from "+
+			"GoroutineManager: %w", err)
+	}
```
related to my comment above: from what I can tell, the only error that `gm.Go(..)` will _ever_ return is `fn.ErrStopping` or `nil`. So to me, it returning this is not actually an error but more just a "state transition" we should be aware of. Which I think is a point towards handling this explicitly in the actual callback passed to `Go` via a quit channel as mentioned above. 
If we do want some idea of if the goroutine manager did its thing from outside the call-back (cause it could also be that the call-back never gets called), then I think a simple `bool` could to the trick since it basically just communicates "handled/not handled due to shutdown"

starius replied:
This is now fixed in new GoroutineManager version. Method `Go` now returns bool (success). Resolving the thread.

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -1627,15 +1645,16 @@ out:
-				defer s.wg.Done()
-
-				if err := s.FlushForwardingEvents(); err != nil {
+			// The error of Go is ignored: if it is shutting down,
+			// the loop will terminate on the next iteration, in
+			// s.gm.Done case.
```
another point for the above case :) 

This comment shows that the caller of `Go` should know that any error returned is just a shutting down error. I dont think the caller should be expected to know that.

ellemouton reviewed (commented):
Thanks for the updates @starius!

Logic looks good, but I have some opinions about the API of the `fn.Go` call that I think is worth discussing before we merge. Would love to hear what @yyforyongyu & @ProofOfKeags think too.

ellemouton commented:
Hmmm, I also want to point out that I think this race described in `fn/goroutine_manager.go` is not actually guarded against:

The `g.mu` mutex _does_ wrap the call to `g.wg.Add(1)` but the use of the `g.mu` in `Stop` does not wrap the call to `g.wg.Wait()`. so Add and Wait can very much still be called at the same time. 

```
// Go starts a new goroutine if the manager is not stopping.
func (g *GoroutineManager) Go(f func(ctx context.Context)) error {
	// Calling wg.Add(1) and wg.Wait() when wg's counter is 0 is a race
	// condition, since it is not clear should Wait() block or not. This
	// kind of race condition is detected by Go runtime and results in a
	// crash if running with `-race`. To prevent this, whole Go method is
	// protected with a mutex. The call to wg.Wait() inside Stop() can still
	// run in parallel with Go, but in that case g.ctx is in expired state,
	// because cancel() was called in Stop, so Go returns before wg.Add(1)
	// call.
	g.mu.Lock()
	defer g.mu.Unlock()

	if g.ctx.Err() != nil {
		return ErrStopping
	}

	g.wg.Add(1)
	go func() {
		defer g.wg.Done()
		f(g.ctx)
	}()

	return nil
}

// Stop prevents new goroutines from being added and waits for all running
// goroutines to finish.
func (g *GoroutineManager) Stop() {
	g.mu.Lock()
	g.cancel()
	g.mu.Unlock()

	// Wait for all goroutines to finish. Note that this wg.Wait() call is
	// safe, since it can't run in parallel with wg.Add(1) call in Go, since
	// we just cancelled the context and even if Go call starts running here
	// after acquiring the mutex, it would see that the context has expired
	// and return ErrStopping instead of calling wg.Add(1).
	g.wg.Wait()
}
```

ProofOfKeags commented:
What's the prio on this? I want to review but I need to balance with other stuff.

yyforyongyu commented on htlcswitch/switch.go:
```diff
@@ -245,8 +246,14 @@ type Switch struct {
 
-	wg   sync.WaitGroup
-	quit chan struct{}
+	// TODO(yy): remove handleLocalResponseWG, once handleLocalResponse runs
```
hmm why it's my TODO😂

starius replied:
`handleLocalResponseWG` was added because `handleLocalResponse` is called in a goroutine, which can't be tracked using GoroutineManagaer. There is an existing [TODO(yy)](https://github.com/lightningnetwork/lnd/blob/c0e85d36ae0263f8b7283f8d0dd5e9f127f760f3/htlcswitch/switch.go#L3022-L3024) to remove the goroutine running `handleLocalResponse`. I copy-pasted that TODO here, since if that TODO is fixed, then handleLocalResponseWG is not needed, so this TODO is also fixed :-)

starius replied:
I [removed](https://github.com/lightningnetwork/lnd/pull/9140#discussion_r1851225183) the special treatment of `handleLocalResponse`, along with the associated WaitGroup and TODOs. It is now managed entirely by GoroutineManager.

Now the PR is not adding new TODOs for @yyforyongyu :-)
Resolving the thread.

yyforyongyu reviewed (commented):
Sorry a bit late in the game, but is there an issue page describing what the issue is?

I also don't understand the struct `GoroutineManager` - it looks like it's putting a mutex to guard the wait group operations?

My instinct is this is solving the wrong problem - we should always know when/where we call `wg.Add` and `wg.Wait`, if not, we should refactor our code so we always know when we cal `wg.Add` and `wg.Wait`. I guess other people have run into this [issue before too](https://github.com/golang/go/issues/23842).

lightninglabs-deploy commented:
@proofofkeags: review reminder
@crypt-iq: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@starius, remember to re-request review from reviewers when ready

ProofOfKeags commented on htlcswitch/switch.go:
```diff
@@ -245,8 +246,14 @@ type Switch struct {
+	// TODO(yy): remove handleLocalResponseWG, once handleLocalResponse runs
+	// without a goroutine. Currently we can't run handleLocalResponse in
+	// gm, since if gm is stopping, the goroutine won't start and it is
+	// unclear if it safe to skip handleLocalResponse.
+	handleLocalResponseWG sync.WaitGroup
```
I don't believe this is necessary as I believe that the composition of waitgroups is equivalent to the waitgroup of the composition of threads when the wait conditions are always called in conjunction with one another.

starius replied:
We discussed this offline.

The main reason for the special handling of `handleLocalResponse` was that, in the old version, it was launched unconditionally (even when the switch was stopping), and switching to `GoroutineManager` introduces a change in behavior. However, we need to ensure that the effects are idempotent to prevent inconsistent states in the event of power failures. If this requirement is met, the behavior change should not pose an issue.

Therefore, I removed the special treatment of `handleLocalResponse`, along with the associated `WaitGroup` and TODOs. It is now managed entirely by `GoroutineManager`.

starius commented:
> Sorry a bit late in the game, but is there an issue page describing what the issue is?

@yyforyongyu Thank you for the suggestion!
I opened https://github.com/lightningnetwork/lnd/issues/9308 to describe the original issue.

> I also don't understand the struct `GoroutineManager` - it looks like it's putting a mutex to guard the wait group operations?

`WaitGroup` cannot be directly used to track goroutines in the scenario we encounter in `htlcswitch`. The issue arises when we have a long-lived object (the switch) that launches goroutines during its lifecycle (via the `GetAttemptResult` method, which calls `wg.Add(1)`) and a `Stop()` method, which cancels running goroutines (using context cancellation) and waits for them to complete (via `wg.Wait()`).  

In this setup, `wg.Add(1)` and `wg.Wait()` can be called in parallel when the `WaitGroup` counter is at 0. At this point, `WaitGroup` cannot determine whether it should wait or not because the outcome depends on the timing and order of these calls. Essentially, this creates a situation where `switch.Stop()` doesn’t know whether to wait for a goroutine launched by `GetAttemptResult` if it was initiated at the same time `Stop()` was called. This results in a race condition.  

`GoroutineManager` resolves this issue by introducing a `Go` method that is synchronized with the `Stop` method. This ensures that either a goroutine is successfully launched or the `Go` method returns `false`. This synchronization is achieved by using a mutex alongside the `WaitGroup`.  

> My instinct is this is solving the wrong problem - we should always know when/where we call `wg.Add` and `wg.Wait`, if not, we should refactor our code so we always know when we cal `wg.Add` and `wg.Wait`. I guess other people have run into this [issue before too](https://github.com/golang/go/issues/23842).

I agree that, ideally, the code should be refactored into an event-loop style, centralizing all goroutine launches and state changes within a single goroutine and using channels to transmit data to and from it. This approach aligns with the patterns we follow in other packages. However, implementing such a change would require significant time and extensive modifications to the package. What are your thoughts?

ellemouton reviewed (commented):
@starius  - I think this still needs to be updated to point to the latest version of `fn` (https://github.com/lightningnetwork/lnd/pull/9270). 

Also - I think you can go ahead and squash in that final commit

starius commented:
I squashed the last commit (https://github.com/lightningnetwork/lnd/pull/9140/commits/deeacc62478bbe5f6b8fe1f2ee6cae02e3d7c8c1), rebased and used GoroutineManager from fn v2. Fortunately fn v1 and fn v2 can be used simultaneously!

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@ellemouton: review reminder
@yyforyongyu: review reminder

starius commented on htlcswitch/switch.go:
```diff
@@ -836,7 +847,8 @@ func (s *Switch) logFwdErrs(num *int, wg *sync.WaitGroup, fwdChan chan error) {
 			}
-		case <-s.quit:
+
+		case <-ctx.Done():
```
Thanks! I replaced `ctx.Done()` with `s.gm.Done()` here and also inside a goroutine launched by `GetAttemptResult`.

starius commented on htlcswitch/switch.go:
```diff
@@ -368,8 +370,11 @@ func New(cfg Config, currentHeight uint32) (*Switch, error) {
 		return nil, err
 	}
 
+	gm := fn2.NewGoroutineManager(context.Background())
```
Thanks! I updated `fn` dependency and used new API!

starius commented on htlcswitch/switch.go:
```diff
@@ -704,12 +709,18 @@ func (s *Switch) ForwardPackets(linkQuit <-chan struct{},
+			s.logFwdErrs(ctx, &numSent, &wg, fwdChan)
+		})
+		if !ok {
+			return nil
```
There is the following `case` above:
```go
case <-s.gm.Done():
  return nil
```
(It used to be `case <-s.quit` before the PR.)

We get into the branch in question if gm is stopped right before the `Go` call, because if it was stopped before, the above `case` would trigger and we would return `nil`. My reasoning to return `nil` was that the exact timing of gm stopping should not affect the end result.

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -85,6 +86,9 @@ var (
 	DefaultMaxFeeExposure = lnwire.MilliSatoshi(500_000_000)
+
+	// background is a shortcut for context.Background.
+	background = context.Background()
```
i dont think we should do this. Rather use a `context.TODO()` where needed

ellemouton replied:
if you rebase on top of https://github.com/lightningnetwork/lnd/pull/9344, then we can also add a context guard here and then we only need a single context.TODO() in `Start()`

starius replied:
Fixed.
There are now 3 top-level methods left which use context.TODO() now:
 - Start
 - ForwardPackets
 - GetAttemptResult

Probably they should get a context argument in the future and it will replace the context.TODO().

ellemouton commented on protofsm/state_machine.go:
```diff
@@ -25,6 +25,9 @@ var (
 	// StateMachine that has been asked to Stop.
 	ErrStateMachineShutdown = fmt.Errorf("StateMachine is shutting down")
+
+	// background is a shortcut for context.Background.
+	background = context.Background()
```
we should not do this. 

consider rebasing on top of https://github.com/lightningnetwork/lnd/pull/9342 which handles the bump to the correct `fn` version and handles updating the statemachine to thread contexts through correctly

starius replied:
Done. This commit is not needed now.

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -493,14 +499,11 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 		select {
 		case n = <-nChan:
-		case <-s.quit:
+		case <-s.gm.Done():
```
i think it is not great to refer to `s.gm` from inside a call-back that is called from `s.gm`. Rather just use the `ctx` provided to the callback which will be cancelled when the gm is shutdown (ie, when gm.Done() would have returned anyways)

starius replied:
Fixed

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -1628,15 +1641,16 @@ out:
+			// The error of Go is ignored: if it is shutting down,
+			// the loop will terminate on the next iteration, in
+			// s.gm.Done case.
+			_ = s.gm.Go(background, func(ctx context.Context) {
```
let `htlcForwarder` take a context and pass in a context in there from the goroutine which is starting it

starius replied:
Fixed

ellemouton commented on htlcswitch/switch.go:
```diff
@@ -3020,8 +3042,12 @@ func (s *Switch) handlePacketSettle(packet *htlcPacket) error {
 		// TODO(yy): remove the goroutine and send back the error here.
-		s.wg.Add(1)
-		go s.handleLocalResponse(packet)
+		ok := s.gm.Go(background, func(ctx context.Context) {
```
rather pass in a context to the calling func. Same for all the others

ellemouton reviewed (commented):
let's hold off here until https://github.com/lightningnetwork/lnd/pull/9344 and https://github.com/lightningnetwork/lnd/pull/9342 are merged as those will make things easier here

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

ellemouton commented:
@starius  - those 2 PRs are in now so I think we can continue here

lightninglabs-deploy commented:
@crypt-iq: review reminder
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

ellemouton reviewed (commented):
looks good after squash 🙏 

Let's follow up soon to replace the TODOs

starius commented:
Squashed the commits.

yyforyongyu commented on htlcswitch/switch.go:
```diff
@@ -524,7 +524,11 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
-	}()
+	})
+	// The switch shutting down is signaled by closing the channel.
+	if !ok {
```
Why do we still need this check? Won't the line `<-ctx.Done()` be hit when it's shutting down?

starius replied:
If `GoroutineManager.Stop` is called before the `Go` method (i.e., the switch is in the process of stopping), the `Go` method will return `false` without launching a new goroutine. In such cases, we should perform the same action as if it had stopped after launching the goroutine - specifically, closing `resultChan`. Failing to close the channel and simply returning it could cause the caller to get stuck indefinitely while waiting to receive from the channel.

yyforyongyu commented on htlcswitch/switch.go:
```diff
@@ -704,12 +708,19 @@ func (s *Switch) ForwardPackets(linkQuit <-chan struct{},
+		ok := s.gm.Go(context.TODO(), func(ctx context.Context) {
+			s.logFwdErrs(ctx, &numSent, &wg, fwdChan)
+		})
+		if !ok {
```
same here - why do we need this check? I think `s.logFwdErrs` will listen on `<-s.gm.Done():` and quit?

starius replied:
This is a similar situation. We should handle it the same way if the goroutine manager is stopped before the Go method is executed.

yyforyongyu reviewed (commented):
> I agree that, ideally, the code should be refactored into an event-loop style, centralizing all goroutine launches and state changes within a single goroutine and using channels to transmit data to and from it. This approach aligns with the patterns we follow in other packages. However, implementing such a change would require significant time and extensive modifications to the package. What are your thoughts?

Have you tried the event loop approach? At a glance I think we only need to add a new channel receiver `attemptResultReq` on `Switch` and read it in the main loop `htlcForwader`? Seems doable as the diff is small https://gist.github.com/yyforyongyu/7cf8d2e2586b2c38d197e05315b9d55d

The other approach is simply removing the `wg.Add` - why do we need it or am I missing anything here?
```diff
diff --git a/htlcswitch/switch.go b/htlcswitch/switch.go
index 720625f2c..5e11ce794 100644
--- a/htlcswitch/switch.go
+++ b/htlcswitch/switch.go
@@ -493,10 +493,7 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 	// Since the attempt was known, we can start a goroutine that can
 	// extract the result when it is available, and pass it on to the
 	// caller.
-	s.wg.Add(1)
 	go func() {
-		defer s.wg.Done()
-
 		var n *networkResult
 		select {
 		case n = <-nChan:
@@ -518,12 +515,15 @@ func (s *Switch) GetAttemptResult(attemptID uint64, paymentHash lntypes.Hash,
 		if err != nil {
 			e := fmt.Errorf("unable to extract result: %w", err)
 			log.Error(e)
-			resultChan <- &PaymentResult{
-				Error: e,
-			}
+			fn.SendOrQuit(
+				resultChan, &PaymentResult{
+					Error: e,
+				}, s.quit,
+			)
 			return
 		}
-		resultChan <- result
+
+		fn.SendOrQuit(resultChan, result, s.quit)
 	}()
 
 	return resultChan, nil
```

I think we are more or less on the same page, as we know it's a temporary mitigation to the issue. And I wanna stress again about the wrong usage of `wg.Add(1)`, as explained from this OG [comment](https://github.com/golang/go/issues/23842#issuecomment-365878004).

Or my question is this - now that we have the new `fn.GoroutineManager`, how are we gonna prevent future development from using it to cover the mistake that a `wg.Add` is called inside a goroutine?

starius commented:
@yyforyongyu 

> The other approach is simply removing the wg.Add - why do we need it or am I missing anything here?

If we remove `wg.Add`, we would end up with an untracked goroutine. The purpose of the `WaitGroup` here is to ensure that all started goroutines are properly accounted for and to wait for their completion in `Switch.Stop()`. This guarantees that all goroutines have finished before the switch is turned off.

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

yyforyongyu commented:
I don't think my comments are addressed. I've provided two alternatives that we don't need the `wg.Add`, plus this question,

> Or my question is this - now that we have the new fn.GoroutineManager, how are we gonna prevent future development from using it to cover the mistake that a wg.Add is called inside a goroutine?

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@proofofkeags: review reminder
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

yyforyongyu commented:
!lightninglabs-deploy mute
